### PR TITLE
GreaterThan guard clause for Integers

### PR DESCRIPTION
**Pull request by chouzar**
This pull request is meant to solve Issue #455 meant to add greater than guard clauses for integer types:

```gleam
case 1, 0 {
  x, y if x > y -> True
  _, _ -> False
}
```

Adding the new operator was a very mechanical process :) thanks a lot for the thorough explanation.

### CODE REVIEW

File: src/grammar.lalrpop [hunk 1]
```diff
@@ -300,6 +300,17 @@ ClauseGuard3: UntypedClauseGuard = {
...
+        location: location(s, e),
+        typ: (),
+        left: Box::new(left),
+        right: Box::new(right),
+    },
+
+    ClauseGuard5 => <>,
+}
+
+ClauseGuard5: UntypedClauseGuard = {
```
Comment (chouzar) [inline src/grammar.lalrpop:313]: I'm really clueless about this file but followed the numbering instructions and it worked just fine :) 

Why are names numbered?

File: src/typ.rs [hunk 2]
```diff
@@ -2396,6 +2396,26 @@ fn infer_clause_guard(
...
+            left,
+            right,
+            ..
+        } => {
+            let left = infer_clause_guard(*left, level, env)?;
+            unify(int(), left.typ(), env)
+                .map_err(|e| convert_unify_error(e, left.location()))?;
+            let right = infer_clause_guard(*right, level, env)?;
+            unify(int(), right.typ(), env)
+                .map_err(|e| convert_unify_error(e, right.location()))?;
```
Comment (chouzar) [inline src/typ.rs:2411]: Ok so here we are just checking the integer types. But I'm guessing rust is able to use the same comparison operator `>` `<` for many different types 🤔 

If that's the case would it be possible to make a broader check? Kind of like using a `Rust::Numeric` type or the like and checking for "numeric" + same type.

File: src/grammar.lalrpop (hunk 1 above)
Comment (lpil) [reply src/grammar.lalrpop:313]: It's to do with operator precedence. The parser tries one set of operators, then another, then another, etc. :)

File: src/typ.rs (hunk 2 above)
Comment (lpil) [reply src/typ.rs:2411]: Gleam doesn't have any features like operator overloading or type classes so the `>` operator only works on Ints. For floats we use `>.`

We might add something here later, but this is what we have for now :)

File: src/ast.rs [hunk 3]
```diff
@@ -304,6 +304,13 @@ pub enum ClauseGuard<Type> {
         right: Box<Self>,
     },
 
+    GreaterThan {
```
Comment (lpil) [inline src/ast.rs:307]: Can you name this `GtInt`, to match the `GtInt` node elsewhere in the AST?

File: src/typ.rs (hunk 2 above)
Comment (chouzar) [reply src/typ.rs:2411]: Right, I forget that we're just using Rust to parse valid Gleam.

File: src/ast.rs (hunk 3 above)
Comment (chouzar) [reply src/ast.rs]: Modified symbol!

**Comment by chouzar** [comment]
Should I update the branch from here?
image

**Comment by lpil** [comment]
I prefer not to use that button as it adds a merge commit, something I don't want on master. I'm happy to rebase this in as is :)

Thank you again for you work! This is great

**Comment by lpil** [comment]
Oh! I forgot! Would you mind making another PR to update the CHANGELOG.md file please?

**Comment by chouzar** [comment]
> Oh! I forgot! Would you mind making another PR to update the CHANGELOG.md file please?

Sure! Changelog notes are added once a PR is approved?

And for the rest of operators `>=`, `<`, `<=`, etc., one PR per operator is fine?

**Comment by lpil** [comment]
> Sure! Changelog notes are added once a PR is approved?

I normally try to get them in first but I forgot this time.

> And for the rest of operators >=, <, <=, etc., one PR per operator is fine? Or could they go bundled?

Either or. Might be nice to let someone else do the Float ones if there's interest but I don't really mind :)

**Comment by chouzar** [comment]
Ok! grouping by type sounds sensible :)

**Pull request merged**